src/plugins/data/public/search/expressions/esdsl.ts

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import {
+  KibanaContext,
+  ExpressionFunctionDefinition,
+} from '../../../../../plugins/expressions/public';
+
+import { getSearchService, getUiSettings } from '../../services';
+import { EsRawResponse } from './es_raw_response';
+import { RequestStatistics, RequestAdapter } from '../../../../inspector/common';
+import { IEsSearchResponse } from '../../../common/search/es_search';
+import { buildEsQuery, getEsQueryConfig } from '../../../common/es_query/es_query';
+
+const name = 'esdsl';
+
+type Input = KibanaContext | null;
+type Output = Promise<EsRawResponse>;
+
+interface Arguments {
+  dsl: string;
+}
+
+export type EsdslExpressionFunctionDefinition = ExpressionFunctionDefinition<
+  typeof name,
+  Input,
+  Arguments,
+  Output
+>;
+
+export const esdsl = (): EsdslExpressionFunctionDefinition => ({
+  name,
+  type: 'es_raw_response',
+  inputTypes: ['kibana_context', 'null'],
+  help: i18n.translate('data.search.esdsl.help', {
+    defaultMessage: 'Run elasticsearch request',
+  }),
+  args: {
+    dsl: {
+      types: ['string'],
+      aliases: ['_', 'q', 'query'],
+      help: i18n.translate('data.search.esdsl.q.help', {
+        defaultMessage: 'Query DSL',
+      }),
+      required: true,
+    },
+  },
+  async fn(input, args, { inspectorAdapters, abortSignal }) {
+    const searchService = getSearchService();
+
+    const dsl = JSON.parse(args.dsl);
+    if (!dsl.body) dsl.body = {};
+
+    const dslBody = dsl.body;
+
+    if (input) {
+      const esQueryConfigs = getEsQueryConfig(getUiSettings());
+      const query = buildEsQuery(
+        dslBody.index,
+        input.query || [],
+        input.filters || [],
+        esQueryConfigs
+      );
+
+      if (!dslBody.query) {
+        dslBody.query = query;
+      } else if (!dslBody.query.bool) {
+        dslBody.query.bool = query.bool;
+      } else if (!dslBody.query.bool.filter) {
+        dslBody.query.bool.filter = query.bool.filter;
+      } else {
+        dslBody.query.bool.filter = dslBody.query.bool.filter.concat(query.bool.filter);
+      }
+    }
```

Comment:
I was testing this out, and noticed a few things with this section of code -- it assumes everything is always going to be in a `bool` and that a `filter` will always be an array.

So for example, I can do this filter as an object instead of array and everything works great:
```
esdsl '{"index": "kibana_sample_data_logs", "size": 4, "body": { "_source": false, "query": {"bool": {"filter": {"term": { "machine.os.keyword": "osx"}}}}}}'
| render
```

But if I add `kibana_context` so that this block executes, it fails. In this case because it assumes `dslBody.query.bool.filter` has a `concat` method.

The other case I ran into was if I try doing a query that isn't in a `bool`, `esdsl` sends a malformed query to ES:
```
kibana_context
| esdsl '{"index": "kibana_sample_data_logs", "size": 4, "body": { "_source": false, "query": { "term": { "machine.os.keyword": "osx"}}}}'
| render
```

The above returns a 400 but works without `kibana_context`. This is the error that comes back:
```
{"statusCode":400,"error":"Bad Request","message":"[parsing_exception] [term] malformed query, expected [END_OBJECT] but found [FIELD_NAME], with { line=1 & col=63 }","attributes":{"error":{"root_cause":[{"type":"parsing_exception","reason":"[term] malformed query, expected [END_OBJECT] but found [FIELD_NAME]","line":1,"col":63}],"type":"parsing_exception","reason":"[term] malformed query, expected [END_OBJECT] but found [FIELD_NAME]","line":1,"col":63}}}
```

It might be good to add a couple tests to verify that these input overrides are working correctly and that valid queries are being sent... Currently all of the tests are just using `{}` as the dsl argument and could probably catch some of these edge cases.

Reply:
ok, so yes, i was assuming that this always produces a `query.bool.filter`:
```
const query = buildEsQuery(
        dslBody.index,
        input.query || [],
        input.filters || [],
        esQueryConfigs
      );
```

also i was assuming that if `dslBody.query.bool.filter` exists that its a array.

so running kibana_context without any arguments will return input with query: [] and filters: [], so it should be exactly the same as running it without it :| ... ah ok not really as in your first example input is undefined.

anyway, input provided and your dsl we get to this: (which is invalid query)
`{"_source":false,"query":{"term":{"machine.os.keyword":"osx"},"bool":{"must":[],"filter":[],"should":[],"must_not":[]}}}`

any idea how to handle this in general enough manner ?

Reply:
i think i figured it out